File: backend/harmony-service/src/main/kotlin/com/chordmind/harmony/service/QuizService.kt
package com.chordmind.harmony.service

import com.chordmind.harmony.client.AIClient
import com.chordmind.harmony.domain.QuizQuestion
import com.chordmind.harmony.domain.QuizResult
import com.chordmind.harmony.domain.QuizType
import com.chordmind.harmony.dto.*
import com.chordmind.harmony.repository.QuizQuestionRepository
import com.chordmind.harmony.repository.QuizResultRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.time.LocalDateTime

@Service
class QuizService(
    private val quizQuestionRepository: QuizQuestionRepository,
    private val quizResultRepository: QuizResultRepository,
    private val aiClient: AIClient
) {
    @Transactional(readOnly = true)
    fun getRandomQuestions(type: QuizType, count: Int): List<QuizQuestion> {
        return quizQuestionRepository.findByType(type).shuffled().take(count)
    }

    @Transactional(readOnly = true)
    fun checkAnswer(request: QuizAnswerRequest): QuizAnswerResult {
        val question = quizQuestionRepository.findById(request.questionId).orElse(null)
            ?: return QuizAnswerResult(request.questionId, false, null)
        
        val correct = question.answer.equals(request.selected, ignoreCase = true)
        
        // Python AI 서비스를 통한 개인화된 피드백 생성
        val aiExplanation = aiClient.generatePersonalizedFeedback(
            userId = request.userId ?: 1L, // 기본값 설정 (실제로는 세션에서 가져와야 함)
            questionType = question.type.name,
            userAnswer = request.selected,
            correctAnswer = question.answer,
            isCorrect = correct,
            previousAnswers = emptyList() // 향후 사용자 히스토리 추가 가능
        )
        
        return QuizAnswerResult(
            questionId = question.id,
            correct = correct,
            explanation = aiExplanation
        )
    }

    @Transactional
    fun saveQuizResult(request: QuizResultRequest): QuizResultResponse {
        val question = quizQuestionRepository.findById(request.questionId).orElseThrow {
            IllegalArgumentException("문제를 찾을 수 없습니다: ${request.questionId}")
        }
        
        val correct = question.answer.equals(request.selected, ignoreCase = true)
        
        val quizResult = QuizResult(
            userId = request.userId,
            question = question,
            selected = request.selected,
            correct = correct
        )
        
        val savedResult = quizResultRepository.save(quizResult)
        
        return QuizResultResponse(
            id = savedResult.id,
            userId = savedResult.userId,
            questionId = savedResult.question.id,
            selected = savedResult.selected,
            correct = savedResult.correct,
            answeredAt = savedResult.answeredAt
        )
    }

    @Transactional(readOnly = true)
    fun getQuizRankings(fromDate: LocalDate, toDate: LocalDate): List<QuizRankingDto> {
        val startDateTime = fromDate.atStartOfDay()
        val endDateTime = toDate.atTime(23, 59, 59)
        
        val rankingData = quizResultRepository.findRankingData(startDateTime, endDateTime)
        
        return rankingData.mapIndexed { index, data ->
            val userId = data[0] as Long
            val totalQuestions = (data[1] as Number).toInt()
            val correctAnswers = (data[2] as Number).toInt()
            val accuracy = if (totalQuestions > 0) {
                (correctAnswers.toDouble() / totalQuestions.toDouble()) * 100
            } else 0.0
            
            QuizRankingDto(
                userId = userId,
                userName = "사용자${userId}", // 실제로는 User Service에서 사용자 이름을 가져와야 함
                score = correctAnswers * 10, // 간단한 점수 계산 로직
                accuracy = accuracy,
                totalQuestions = totalQuestions,
                correctAnswers = correctAnswers,
                rank = index + 1
            )
        }
    }
} 

File: frontend/src/services/quizService.ts
import { apiClient, ApiResponse } from './apiClient';

export type QuizType = 'CHORD_NAME' | 'PROGRESSION' | 'INTERVAL' | 'SCALE';

export interface QuizQuestion {
  id: number;
  type: QuizType;
  questionText: string;
  imageUrl?: string | null;
  choices: QuizChoice[];
  correctAnswer: string;
  explanation?: string | null;
  difficulty: number;
}

export interface QuizChoice {
  id: number;
  choiceText: string;
  isCorrect: boolean;
}

export interface QuizAnswerRequest {
  questionId: number;
  selected: string;
}

export interface QuizAnswerResult {
  questionId: number;
  correct: boolean;
  explanation?: string | null;
}

export interface QuizResultRequest {
  userId: number;
  questionId: number;
  selected: string;
}

export interface QuizResultResponse {
  id: number;
  userId: number;
  questionId: number;
  selected: string;
  correct: boolean;
  answeredAt: string;
}

export interface QuizRankingDto {
  userId: number;
  score: number;
}

class QuizService {
  // 랜덤 퀴즈 문제 조회
  async fetchQuizQuestions(type: QuizType = 'CHORD_NAME', count = 5): Promise<ApiResponse<QuizQuestion[]>> {
    const params = apiClient.buildQueryParams({ type, count });
    return apiClient.get<QuizQuestion[]>(`/api/harmony-quiz/random?${params}`);
  }

  // 퀴즈 답안 제출 및 채점
  async submitQuizAnswer(data: QuizAnswerRequest): Promise<ApiResponse<QuizAnswerResult>> {
    return apiClient.post<QuizAnswerResult>('/api/harmony-quiz/answer', data);
  }

  // 퀴즈 결과 저장
  async saveQuizResult(request: QuizResultRequest): Promise<ApiResponse<QuizResultResponse>> {
    return apiClient.post<QuizResultResponse>('/api/harmony-quiz/result', request);
  }

  // 퀴즈 랭킹 조회
  async getQuizRankings(from: string, to: string): Promise<ApiResponse<QuizRankingDto[]>> {
    const params = apiClient.buildQueryParams({ from, to });
    return apiClient.get<QuizRankingDto[]>(`/api/harmony-quiz/rankings?${params}`);
  }
}

export const quizService = new QuizService(); 